File: 1612-check-if-two-expression-trees-are-equivalent/1612-check-if-two-expression-trees-are-equivalent.cpp
/**
 * Definition for a binary tree node.
 * struct Node {
 *     char val;
 *     Node *left;
 *     Node *right;
 *     Node() : val(' '), left(nullptr), right(nullptr) {}
 *     Node(char x) : val(x), left(nullptr), right(nullptr) {}
 *     Node(char x, Node *left, Node *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void dfs(Node* node,vector<int>& freq){
        //Base Case
        if(!node) return;
        //check if node val is not + if not then increment the corresponding mapping index int freq.
        if(node->val != '+') freq[node->val-'a']++;
        //Frequency calculated by traversing the whole tree by left and right recursive call
        dfs(node->left,freq);
        dfs(node->right,freq);
    }
    bool checkEquivalence(Node* root1, Node* root2) {
        //get freq filled
        vector<int> freq(26,0);
        dfs(root1,freq);
        dfs(root2,freq);
        //check of any freq is odd if yes return false else return true
        for(int i=0;i<26;i++) {
            if(freq[i]%2 != 0) return false;
        }
        return true;
    }
};

File: 1612-check-if-two-expression-trees-are-equivalent/1612-check-if-two-expression-trees-are-equivalent.java
/**
 * Definition for a binary tree node.
 * class Node {
 *     char val;
 *     Node left;
 *     Node right;
 *     Node() {this.val = ' ';}
 *     Node(char val) { this.val = val; }
 *     Node(char val, Node left, Node right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void dfs(Node node,int[] freq) {
        //Base Case
        if(node==null) return;
        //condition to check if i dont have a + in the node
        if(node.val != '+') freq[node.val-'a']++;
        //recurse to left and right side to cover the whole tree
        dfs(node.left,freq);
        dfs(node.right,freq);
        
    }
    public boolean checkEquivalence(Node root1, Node root2) {
        int[] freq = new int[26];
        dfs(root1,freq);
        dfs(root2,freq);
        //if any freq is odd return false else return true;
        for(int i=0;i<26;i++){
            if(freq[i]%2!=0) return false;
        }
        return true;
    }
}